pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,418 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identitybackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	k8sTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "crd-allocator")
+)
+
+func NewCRDBackend(c CRDBackendConfiguration) (allocator.Backend, error) {
+	return &crdBackend{CRDBackendConfiguration: c}, nil
+}
+
+type CRDBackendConfiguration struct {
+	NodeName string
+	Store    cache.Store
+	Client   clientset.Interface
+	KeyType  allocator.AllocatorKey
+}
+
+type crdBackend struct {
+	CRDBackendConfiguration
+}
+
+func (c *crdBackend) DeleteAllKeys() {
+}
+
+// stripK8sPrefix strips the 'k8s:' prefix in the labels generated by
+// AllocatorKey.GetAsMap (when the key is k8s labels). In the CRD identity case
+// we map the labels directly to the ciliumidentity CRD instance, and
+// kubernetes does not allow ':' in the name of the label.
+func stripK8sPrefix(old map[string]string) map[string]string {
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		k = strings.TrimPrefix(k, "k8s:")
+		fixup[k] = v
+	}
+	return fixup
+}
+
+// AllocateID will create an identity CRD, thus creating the identity for this
+// key-> ID mapping.
+// Note: This does not create a reference to this node to indicate that it is
+// using this identity. That must be done with AcquireReference.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) error {
+	identity := &v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   id.String(),
+			Labels: stripK8sPrefix(key.GetAsMap()),
```

Comment:
I remember putting this in to get things going and start testing but for it to be reliable, do we instead need to translate `:` into another character? We can still get non `k8s:` prefixes, right? Which would then break the identity allocation.

Reply:
We do need to translate the `:`. I started translating it but when running in k8s I expected that they would always be `k8s:` sources. aanm suggested that we may also have `container:` sources. I'll check this again.

Reply:
Ok, from what I can tell, it is unlikely to get non-`k8s:` prefixed labels here. The problem is whether we care how the labels looks in k8s:
```
 $ kubectl get ciliumid 26636 -o json
{
    "apiVersion": "cilium.io/v2",
    "kind": "CiliumIdentity",
    "metadata": {
        "creationTimestamp": "2019-07-15T20:36:52Z",
        "generation": 1,
        "labels": {
            "class": "xwing",
            "io.cilium.k8s.policy.cluster": "default",
            "io.cilium.k8s.policy.serviceaccount": "default",
            "io.kubernetes.pod.namespace": "default",
            "org": "alliance"
        },
        "name": "26636",
        "resourceVersion": "10253",
        "selfLink": "/apis/cilium.io/v2/ciliumidentities/26636",
        "uid": "a6ddb730-8d80-4d6f-97ff-f6f03db393d9"
    },
    "status": {
        "nodes": {
            "192.168.34.11": "2019-07-15T21:23:54Z"
        }
    }
}
```

The only other separator character allowed, which I translated `:` before, is `_`. The labels would look like:
```
        "labels": {
            "k8s_class": "xwing",
            "k8s_io.cilium.k8s.policy.cluster": "default",
            "k8s_io.cilium.k8s.policy.serviceaccount": "default",
            "k8s_io.kubernetes.pod.namespace": "default",
            "k8s_org": "alliance"
        },
```

I'll think about this some more, but I'll leave it as is right now.